drivers/dac/dac_max22017.c

```diff
@@ -0,0 +1,636 @@
+#define MAX22017_AO_STA_BUSY_STA GENMASK(15, 14)
+#define MAX22017_AO_STA_SLEW_STA GENMASK(13, 12)
+#define MAX22017_AO_STA_FAIL_STA BIT(0)
+
+#define MAX22017_AO_CNFG_OFF                  0x42
+#define MAX22017_AO_CNFG_AO_LD_CNFG           GENMASK(15, 14)
+#define MAX22017_AO_CNFG_AO_CM_SENSE          GENMASK(13, 12)
+#define MAX22017_AO_CNFG_AO_UNI               GENMASK(11, 10)
+#define MAX22017_AO_CNFG_AO_MODE              GENMASK(9, 8)
+#define MAX22017_AO_CNFG_AO_OPENWIRE_DTCT_LMT GENMASK(5, 4)
+#define MAX22017_AO_CNFG_AI_EN                GENMASK(3, 2)
+#define MAX22017_AO_CNFG_AO_EN                GENMASK(1, 0)
+
+#define MAX22017_AO_SLEW_RATE_CHn_OFF(n)                (0x43 + n)
+#define MAX22017_AO_SLEW_RATE_CHn_AO_SR_EN_CHn          BIT(5)
+#define MAX22017_AO_SLEW_RATE_CHn_AO_SR_SEL_CHn         BIT(4)
+#define MAX22017_AO_SLEW_RATE_CHn_AO_SR_STEP_SIZE_CHn   GENMASK(3, 2)
+#define MAX22017_AO_SLEW_RATE_CHn_AO_SR_UPDATE_RATE_CHn GENMASK(1, 0)
+
+#define MAX22017_AO_DATA_CHn_OFF(n)     (0x45 + n)
+#define MAX22017_AO_DATA_CHn_AO_DATA_CH GENMASK(15, 0)
+
+#define MAX22017_AO_OFFSET_CORR_CHn_OFF(n)       (0x47 + (2 * n))
+#define MAX22017_AO_OFFSET_CORR_CHn_AO_OFFSET_CH GENMASK(15, 0)
+
+#define MAX22017_AO_GAIN_CORR_CHn_OFF(n)     (0x48 + (2 * n))
+#define MAX22017_AO_GAIN_CORR_CHn_AO_GAIN_CH GENMASK(15, 0)
+
+#define MAX22017_SPI_TRANS_ADDR    GENMASK(7, 1)
+#define MAX22017_SPI_TRANS_DIR     BIT(0)
+#define MAX22017_SPI_TRANS_PAYLOAD GENMASK(15, 0)
+
+LOG_MODULE_REGISTER(dac_max22017, CONFIG_DAC_LOG_LEVEL);
+
+struct max22017_config {
+	struct spi_dt_spec spi;
+	uint8_t resolution;
+	uint8_t nchannels;
+	const struct gpio_dt_spec gpio_reset;
+	const struct gpio_dt_spec gpio_ldac;
+	const struct gpio_dt_spec gpio_busy;
+	const struct gpio_dt_spec gpio_int;
+	uint8_t latch_mode[MAX22017_MAX_CHANNEL];
+	uint8_t polarity_mode[MAX22017_MAX_CHANNEL];
+	uint8_t dac_mode[MAX22017_MAX_CHANNEL];
+	uint8_t ovc_mode[MAX22017_MAX_CHANNEL];
+	uint16_t timeout;
+	bool crc_mode;
+};
+
+struct max22017_data {
+	const struct device *dev;
+	struct k_sem lock;
+	struct k_work int_work;
+	struct gpio_callback callback_int;
+	bool crc_enabled;
+};
+
+static int max22017_reg_read(const struct device *dev, uint8_t addr, uint32_t *value)
+{
+	int ret;
+	const struct max22017_config *config = dev->config;
+	const struct max22017_data *data = dev->data;
+	size_t recv_len = 3;
+
+	*value = 0;
+	addr = FIELD_PREP(MAX22017_SPI_TRANS_ADDR, addr) | FIELD_PREP(MAX22017_SPI_TRANS_DIR, 1);
+
+	if (data->crc_enabled) {
+		recv_len += 1;
+	}
+
+	const struct spi_buf txb[] = {
+		{
+			.buf = &addr,
+			.len = 1,
+		},
+	};
+	const struct spi_buf rxb[] = {
+		{
+			.buf = value,
+			.len = recv_len,
+		},
+	};
+
+	struct spi_buf_set tx = {
+		.buffers = txb,
+		.count = ARRAY_SIZE(txb),
+	};
+
+	struct spi_buf_set rx = {
+		.buffers = rxb,
+		.count = ARRAY_SIZE(rxb),
+	};
+
+	ret = spi_transceive_dt(&config->spi, &tx, &rx);
+
+	if (!data->crc_enabled) {
+		return ret;
```

Comment:
I might have been confused by the spi_transceive_dt() API, but reading 2 bytes into the 3 bytes buffer makes the value aligned anyway.

I was expecting to read 2 bytes into the rx buffer, but when doing that the clock stops after the first byte of data has been sent on MISO

Reply:
So are you saying that if you set `rx.count` to 2, you can only read a single byte and after that the clock stops?

Reply:
Yes, which is a bit surprising to me.
I have to send a 1 byte tx (MOSI) to read the register and then read (MISO) the 2 bytes reply (3 bytes with CRC).

Experimentally, setting rx.count to 2 gets me only the first byte of reply (clock/enable stops after the first byte is received).

Looks like the rx.count starts at the same time as the tx is starting?
I find this strange, but as it is my first ever Zephyr driver, I might have missed something?

I didn't dig too much into the SPI implementation of the board I'm using to verify this is the expected behavior (SDP-k1 from ADI : https://www.analog.com/en/resources/evaluation-hardware-and-software/evaluation-boards-kits/sdp-k1.html)

Reply:
I have checked several driver implementations, and I think what you observed is the correct behavior. Since SPI is full duplex, it sends and receives at the same time. So, if you set `tx.count` to 1 and `rx.count` to 2 you will actually observe 2 clock cycles because the RX starts at the same time TX start is starting as you suspect. In this case you need at least 3 clock cycles to get the whole reply packet (4 if crc is enabled).

Reply:
I am still confused about the endianness though. After the transaction you should have something like
```c
((uint8_t *) &value)[0] = 0x00; // redundant byte received during TX
((uint8_t *) &value)[1] = 0xXX; // Payload MSB
((uint8_t *) &value)[2] = 0xYY; // Payload LSB
```
on a little endian host. So, it seems you still need to do a shift and conversion.

Did you compare the values received in CRC and non-CRC modes?

Reply:
@ttmut : thx for helping me checking the spi_transceive_dt() behaviour.

regarding the endianness, I've reworked the function to use a temporary uint8_t[4] buffer instead of using a larger uint32_t as input and messing with it...
The code now has the same conversion for CRC and non CRC transfers and it avoids passing an uint32_t to reg_read().

Hope the code is clearer/cleaner now.